### ISSUE TITLE
args.train_batch_size incorrectly modified for run_squad.py

### ISSUE THREAD

**Issue by FruVirus** [NVIDIA/DeepLearningExamples#454]
1. The line below modifies `args._train_batch_size` using `args.gradient_accumulation_steps`. 

https://github.com/NVIDIA/DeepLearningExamples/blob/1cad1801646dee0e96e344b43c796181d6f05564/PyTorch/LanguageModeling/BERT/run_squad.py#L874

However, later on, `args.train_batch_size` is again divided by `args.gradient_accumulation_steps` here: 

https://github.com/NVIDIA/DeepLearningExamples/blob/c742ed42dc8eabaf75fc5388c023e6de37a61008/PyTorch/LanguageModeling/BERT/run_squad.py#L910

2. In addition, `args.train_batch_size` is then multiplied by the number of gpus here:

https://github.com/NVIDIA/DeepLearningExamples/blob/c742ed42dc8eabaf75fc5388c023e6de37a61008/PyTorch/LanguageModeling/BERT/run_squad.py#L1024

This would result in an incorrect batch size if `args.gradient_accumulation_steps` != 1.

**Comment by sharathts** [comment]
n_gpu is a variable set when using torch DP. When running with torch DDP, it is always set to 1. 

So dataloader multiplying by n_gpu is the right thing to do. (Because when using torch.DP,  RandomSampler is used and the batch size across all gpus has to be specified)

Train batch size being divided by gradient accumulation steps is intended. See scripts/pretraining.sh for expected usage and behavior.

**Comment by FruVirus** [comment]
I understand the setting of batch size between DP and DDP.

However, args.train_batch_size is divided by args.gradient_accumulation_steps twice in run_squad.py---once on line 879 and again on line 915. On line 879, args.train_batch_size is updated to its new value. So on line 915, dividing again would be incorrect?

the run_squad.sh script doesnt do anything special with the batch size argument besides assignment.

**Comment by sharathts** [comment]
Thank for pointing it out, it will be fixed soon. Feel free to submit a PR as well.

**Comment by sharathts** [comment]
The division in https://github.com/NVIDIA/DeepLearningExamples/blob/c742ed42dc8eabaf75fc5388c023e6de37a61008/PyTorch/LanguageModeling/BERT/run_squad.py#L910 is to compute the number of training steps. Since batch size has already been scaled down, the number of training steps have to be scaled down by accumulation steps.

With 20 training samples,  train_batch_size = 4, gradient accumulation steps = 2, num_epochs=1:
Code changes train_batch_size to 4/2 = 2

num_train_steps = 20 / 2 / 2 * 1 = 5, which is correct
If you don't divide by accumulation steps, you end up with num_train_steps = 20 / 2 * 1 = 10, which with a train_batch_size(accumulated batch size) of 4 is incorrect.

**Comment by FruVirus** [comment]
with the above way of calculating number of training steps when accumulation steps > 1, is the setting of `batch_size` for the `DataLoader` still correct? At this point, `args.train_batch_size` is 4/2 = 2.

https://github.com/NVIDIA/DeepLearningExamples/blob/c742ed42dc8eabaf75fc5388c023e6de37a61008/PyTorch/LanguageModeling/BERT/run_squad.py#L1024
